contrib/generic_proxy/filters/network/source/proxy.cc

```diff
@@ -225,15 +393,49 @@ void Filter::resetStreamsForUnexpectedError() {
   }
 }
 
+void Filter::closeDownstreamConnection() {
+  if (downstream_connection_closed_) {
+    return;
+  }
+  downstream_connection_closed_ = true;
+  connection().close(Network::ConnectionCloseType::FlushWrite);
+}
+
 void Filter::mayBeDrainClose() {
-  if (drain_decision_.drainClose() && active_streams_.empty()) {
+  if ((drain_decision_.drainClose() || stream_drain_decision_) && active_streams_.empty()) {
     onDrainCloseAndNoActiveStreams();
   }
 }
 
 // Default implementation for connection draining.
-void Filter::onDrainCloseAndNoActiveStreams() {
-  connection().close(Network::ConnectionCloseType::FlushWrite);
+void Filter::onDrainCloseAndNoActiveStreams() { closeDownstreamConnection(); }
+
+void Filter::bindUpstreamConn(Upstream::TcpPoolData&& tcp_pool_data) {
+  ASSERT(config_->codecFactory().protocolOptions().bindUpstreamConnection());
+  upstream_manager_ = std::make_unique<UpstreamManagerImpl>(*this, std::move(tcp_pool_data));
```

Comment:
As the comment here https://github.com/envoyproxy/envoy/pull/26618/files#diff-b958247f89d87f025a29ad03f02d568e8c37cc05f7cc8890356692f6c0543b49R140, maybe worth a assertion here `ASSERT(upstream_manager_ == nullptr)`

Reply:
got it.